showPicker() method of inputs doesn't trigger datalist suggestions

lukewarlow opened the pull request:
#### dddc5bb328a4b8636946a735281b09101f1d3161
<pre>
showPicker() method of inputs doesn&apos;t trigger datalist suggestions
<a href="https://bugs.webkit.org/show_bug.cgi?id=261701">https://bugs.webkit.org/show_bug.cgi?id=261701</a>

Reviewed by Aditya Keerthi.

Implements showPicker for TextFieldInputs that have a datalist.
This does not work on iOS a follow up will address that.

* LayoutTests/fast/forms/datalist/datalist-show-picker-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-show-picker.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::showPicker):
* Source/WebCore/html/TextFieldInputType.h:

Canonical link: <a href="https://commits.webkit.org/269981@main">https://commits.webkit.org/269981@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7fff9a2a64fdccd0e8bfdffe3909dbb64048937

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26044 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22026 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24175 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24394 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22528 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24143 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1556 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20656 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26635 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21566 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27806 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21907 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21847 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25595 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1259 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18945 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1269 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21338 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5791 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1670 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1591 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.cpp:
```diff
@@ -70,6 +70,9 @@ void WebDataListSuggestionPicker::displayWithActivationType(WebCore::DataListSug
     }
 
+    if (type == WebCore::DataListSuggestionActivationType::ShowPickerCalled)
+        close();
```
Why is this needed?

lukewarlow replied:
You need to close the list before showing it again else it ends up toggling and you end up with a closed datalist if you press the trigger button twice.

lukewarlow replied:
Does that answer the question well enough or would you like me to dig into the exact reason @pxlcoder ?

pxlcoder replied:
I think it might be good to look a bit more into that, since this is a little confusing, and seems like a workaround.

It may have something to do with the fact that we close the list when clicking away from it?

lukewarlow replied:
Yeah it's definitely a workaround I'm not entirely sure on the cause of the issue it's working around.

If it was clicking outside I'm not sure why closing and opening it would fix it.

I'll build it locally again and try and track down the exact code path it goes down and see if I can be fixed a slightly more proper way.

lukewarlow replied:
I've just realised it's this https://bugs.webkit.org/show_bug.cgi?id=261639 the datalist is rendering just in the wrong location. When I do it on my other monitor it works fine. So I'll go ahead and remove this "fix" and it'll just be part of that bug to fix it.

lukewarlow committed:
showPicker() method of inputs doesn't trigger datalist suggestions
https://bugs.webkit.org/show_bug.cgi?id=261701

Reviewed by Aditya Keerthi.

Implements showPicker for TextFieldInputs that have a datalist.
This does not work on iOS a follow up will address that.

* LayoutTests/fast/forms/datalist/datalist-show-picker-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-show-picker.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::showPicker):
* Source/WebCore/html/TextFieldInputType.h:

Canonical link: https://commits.webkit.org/269981@main
